client/src/modules/vouchers/complex-voucher.ctrl.js

```diff
@@ -105,6 +111,134 @@ function ComplexJournalVoucherController(Vouchers, $translate, Currencies, Sessi
       });
   }
 
+  // open generic income function
+  function openGenericIncomeTool() {
+    Toolkit.open(genericIncomeTool)
+      .then(function (result) {
+        if (!result) { return; }
+
+        var rows = result.rows;
+        var n = result.rows.length;
+
+        // initial clear
+        vm.Voucher.clear();
+
+        while (n--) {
```

Comment:
Could this be written as a helper method  on the `VoucherFormService`? It seems like it may be used a lot. 

For example: 

```js 
// voucher-form.service.js
VoucherForm.prototype.replaceFormRows = function replaceFormRows(rows) { 
  this.clear();
  
  rows.forEach(function (row) { 
    //...
  });
}
```

I would also encourage using methods on arrays like `forEach` and `map`, `while` loops are useful for creating utilities that do not create an additional scope however methods like `forEach` and `map` are much easier to read and understand for other developers.